src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs

```diff
@@ -65,18 +65,64 @@ public static ValueTask<QuicConnection> ConnectAsync(QuicClientConnectionOptions
 
         static async ValueTask<QuicConnection> StartConnectAsync(QuicClientConnectionOptions options, CancellationToken cancellationToken)
         {
+            MsQuicSafeHandle? config = MsQuicConfiguration.Create(options);
             QuicConnection connection = new QuicConnection();
+            connection._configuration = config;
 
             using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
 
-            if (options.HandshakeTimeout != Timeout.InfiniteTimeSpan && options.HandshakeTimeout != TimeSpan.Zero)
-            {
-                linkedCts.CancelAfter(options.HandshakeTimeout);
-            }
-
             try
             {
-                await connection.FinishConnectAsync(options, linkedCts.Token).ConfigureAwait(false);
+                if (!options.RemoteEndPoint.TryParse(out string? host, out IPAddress? address, out int port))
+                {
+                    throw new ArgumentException(SR.Format(SR.net_quic_unsupported_endpoint_type, options.RemoteEndPoint.GetType()), nameof(options));
+                }
+
+                if (address is not null && host is null)
+                {
+                    if (options.HandshakeTimeout != Timeout.InfiniteTimeSpan && options.HandshakeTimeout != TimeSpan.Zero)
+                    {
+                        linkedCts.CancelAfter(options.HandshakeTimeout);
```

Comment:
This feels this is wrong. I think the `options.HandshakeTimeout` should cover timeout for `ConnectAsync` itself not for each address we try. In practice the observed timeout would depend on number of addresses given host resolves to. 

But we do not have any real explanation 
https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quicconnectionoptions.handshaketimeout?view=net-9.0

On the other hand, It already has `CancellationToken` to it somewhat feels duplicate.